### DISCUSSION TITLE
custom lsp_interop function for LSP type definitions

### DISCUSSION THREAD

**Discussion by benyn** [nvim-treesitter/nvim-treesitter-textobjects#396] [Q&A]
I love the `peek_definition_code` command and am trying to create a custom function for type definitions. I'm able to get the floating window near the cursor with the following code, but unfortunately it only shows a single line.
```lua
-- "@class.outer" also shows a single line
require("nvim-treesitter.textobjects.lsp_interop").peek_definition_code("@function.outer", nil, "textDocument/typeDefinition")
```
How might I get the full type definition? It'd be even better if `typeDefinition` is officially supported, but if that's not in the roadmap, would love to get some pointers.

**Comment by kiyoon** [comment]
How is it showing a single line? For me running the code you provided shows the entire function in python.

**Reply by benyn** [reply]
I'm trying this in TypeScript. Function type definitions are fine, but `interface`, `type`, `enum`, `namespace`s are not. They just show the first line before the opening curly bracket.

Or, does `typeDefinition` capture only the lines before the opening curly bracket by design? I noticed that, for functions, `typeDefinition` only shows the signature while `definition` shows the full function.
![image](https://user-images.githubusercontent.com/105991837/219737206-83054895-3898-4e6c-ae93-c433b1c02250.png)
![image](https://user-images.githubusercontent.com/105991837/219735100-59d024cf-87c1-4033-a7ca-3942b74a7c1f.png)

**Reply by kiyoon** [reply]
I think you can use `definition` which is what this plugin seems to be using for keymaps set by setup. I don't know much about LSP document types, sorry

**Reply by theHamsta** [reply]
And `require("nvim-treesitter.textobjects.lsp_interop").peek_definition_code("@class.outer", nil, "textDocument/typeDefinition")` doesn't work? LSP mostly just gives you the range of the identifier of the definition. `lsp_interop` will try to construct a textobject if the cursor would be right on the definition. If there is nothing it will just show the current line.

**Reply by benyn** [reply]
I'm not sure if I understand your suggestion. I find `typeDefinition` very useful and I currently look it up via Telescope, but I'd like to use Treesitter style floating window since I don't need the prompt and search result list that come with Telescope.

Is there a way to force `typeDefinition` to pick up the content between the curly brackets, not just the signature line? Not sure why `@function.outer` and `@class.outer` won't work.

**Reply by benyn** [reply]
> And `require("nvim-treesitter.textobjects.lsp_interop").peek_definition_code("@class.outer", nil, "textDocument/typeDefinition")` doesn't work?

No, unfortunately, it doesn't. Those screenshots were taken with `@class.outer`.

**Reply by theHamsta** [reply]
And when you don't peek but instead go to the type definition, then `@class.outer` would work? There could just be a missing query.

**Reply by benyn** [reply]
Ah, `@class.outer` doesn't work there, either. `vac` on an actual `class` highlights the entire class, but `vac` on a `type`, `interface` or `enum` doesn't select anything. Is this because TypeScript doesn't support it? I just quickly looked at `ecma/textobjects.scm` (which `typescript/textobjects.scm` refers to) and it doesn't look like there's any other alternatives.

**Reply by benyn** [reply]
`typeDefinition` works in Go files, so looks like it's a TypeScript problem? Found explicit mentions of `struct_type` and `interface_type` in `go/textobjects.scm`.

**Reply by benyn** [reply]
Or, I should say, it's because both TypeScript and JavaScript `textobjects.scm` files simply inherit `ecma` when `TypeScript` has additional syntax around `type`, `interface`, `enum`? Is this something I can manually edit locally? Or should this be supported by TypeScript LSP first?

**Reply by theHamsta** [reply]
No, this has nothing to do with LSP. `typescript` just misses here https://github.com/nvim-treesitter/nvim-treesitter-textobjects/blob/15d9c62cb04079cf440ceb6882f2cbfaed66eda1/queries/typescript/textobjects.scm#L1 queries that define `@class.outer` for typescript.

**Reply by kiyoon** [reply]
@benyn Yeah, sounds like some queries are missing. I think LSP is good but we need to add support for the missing ones. You can manually edit the `textobjects.scm` for typescript to see if it makes any difference

**Reply by theHamsta** [reply]
wow, incredible that it needed five contributors to write `; inherit ecma` :smile: . We changed that a lot

**Reply by benyn** [reply]
I copied the following two lines from Go's `textobjects.scm` and pasted them below `; inherit ecma`, but I got `Vim(append):Error executing lua callback: ...nvim/runtime/lua/vim/treesitter/query.lua:219: query: invalid node type at position 3055 for language typescript`.
```
(type_declaration
  (type_spec (type_identifier) (interface_type) @class.inner)) @class.outer
```
I clearly lack the necessary understanding of query syntax. It sounds like you guys are already stretched thin per the most recent issue calling for maintainers, but shall I go ahead and create an issue for this just to keep track of it?

**Reply by theHamsta** [reply]
Do you know about https://github.com/nvim-treesitter/playground or `:lua vim.treesitter.show_tree()<cr>` when you're on nightly Neovim? You can also select typescript on this webpage an paste some typescript code https://tree-sitter.github.io/tree-sitter/playground (make sure to click on the checkbox "query")

The queries are parser specific and will need to be updated for every language? E.g. when I paste

```ts
interface Logger {
  log: (message: string) => void;
}
```
I can see that the parsing result will be
```
[program](https://tree-sitter.github.io/tree-sitter/playground#) [0, 0] - [4, 0]
  [interface_declaration](https://tree-sitter.github.io/tree-sitter/playground#) [0, 0] - [2, 1]
    name: [type_identifier](https://tree-sitter.github.io/tree-sitter/playground#) [0, 10] - [0, 16]
    body: [object_type](https://tree-sitter.github.io/tree-sitter/playground#) [0, 17] - [2, 1]
      [property_signature](https://tree-sitter.github.io/tree-sitter/playground#) [1, 2] - [1, 32]
        name: [property_identifier](https://tree-sitter.github.io/tree-sitter/playground#) [1, 2] - [1, 5]
        type: [type_annotation](https://tree-sitter.github.io/tree-sitter/playground#) [1, 5] - [1, 32]
          [function_type](https://tree-sitter.github.io/tree-sitter/playground#) [1, 7] - [1, 32]
            parameters: [formal_parameters](https://tree-sitter.github.io/tree-sitter/playground#) [1, 7] - [1, 24]
              [required_parameter](https://tree-sitter.github.io/tree-sitter/playground#) [1, 8] - [1, 23]
                pattern: [identifier](https://tree-sitter.github.io/tree-sitter/playground#) [1, 8] - [1, 15]
                type: [type_annotation](https://tree-sitter.github.io/tree-sitter/playground#) [1, 15] - [1, 23]
                  [predefined_type](https://tree-sitter.github.io/tree-sitter/playground#) [1, 17] - [1, 23]
            return_type: [predefined_type](https://tree-sitter.github.io/tree-sitter/playground#) [1, 28] - [1, 32]
```
So the following query
```scm

 ; inherits: ecma 
(interface_declaration) @class.outer
```

should be able to select interfaces. 

When digging deeper into the structure of the tree. You can also add `@class.inner` (might need additional work to trim `{` `}`)
```scm
 ; inherits: ecma 
(interface_declaration
  body: (_) @class.inner) @class.outer
```

The Go parser has totally different names for everything because the Go grammar has a different structure.

**Reply by theHamsta** [reply]
This is how it would look on the website. Just experiment with the queries until `@class.outer` selects what you want. Probably `(<node name here>) @class.outer` should be enough.

![grafik](https://user-images.githubusercontent.com/7189118/219819126-1cbc8e9d-8c3e-4cca-a117-8095703f5431.png)

![grafik](https://user-images.githubusercontent.com/7189118/219818816-cf8fe589-3444-4b3d-b628-61da03b94397.png)

**Reply by benyn** [reply]
Yes! That did the trick. Thank you so much. I didn't know about the playground or website. I'll play with it and add some stuff to the `.scm` file. I can also create a PR for a few obvious things like `type`, `namespace` and `enum`, although I'm not sure if what I'd be able to put together would meet your quality standards. Let me know.

**Reply by kiyoon** [reply]
@benyn Thanks for trying out. Queries should be designed for each language and you can't just copy from another language. That's why it requires a lot of work and contribution. There's [playground](https://github.com/nvim-treesitter/playground) that lets you play with and design queries. If you're interested you can also play with it. I might take a look but also I don't know really anything about typescript

**Reply by kiyoon** [reply]
@theHamsta Oh wow, thanks for this. I wasn't following the conversation to the latest because I got distracted by something else. Does the nightly playground also provide you to design query or does it only show the tree?

**Reply by theHamsta** [reply]
@benyn please consider contributing any findings for typescript back to this repo as it will help others!

**Reply by benyn** [reply]
Will do! Thank you both for this incredible help.

**Reply by kiyoon** [reply]
@benyn Yes contribution is welcome, and I'm glad you enjoy this. For me I found [this TJ's video](https://www.youtube.com/watch?v=v3o9YaHBM4Q) helpful to understand treesitter queries and what it can do.

**Reply by benyn** [reply]
Quick question: per @theHamsta 's comment, I'm trying to trim `{` and `}` from `@class.inner`. I modeled the query after the ones in `ecma` for functions, but for some reason, the last semicolon `;` is left out as you can see in the screenshot. Any idea how to address this?
```
(interface_declaration
  body: (object_type . "{" . (_) @_start @_end (_)? @_end . "}"
 (#make-range! "class.inner" @_start @_end)))
```
![Screenshot 2023-02-17 at 4 56 50 PM](https://user-images.githubusercontent.com/105991837/219822839-566ff44e-6879-4103-a614-6fcd91b2d73d.png)

**Reply by theHamsta** [reply]
replace `(_)` by `_`

**Reply by theHamsta** [reply]
`(_)` only matches named nodes. Press `a` in the playground to see anonymous nodes which are hidden by default

**Reply by benyn** [reply]
Like this? Still doesn't work. :(
```
(interface_declaration
  body: (object_type . "{" . _ @_start @_end _? @_end . "}"
 (#make-range! "class.inner" @_start @_end)))
```

**Reply by benyn** [reply]
I submitted a PR for this noting the `;` issue. Let me know what you think!

**Reply by theHamsta** [reply]
```
(interface_declaration
  body: (object_type . "{" . _ @_start _? @_end . "}"
 (#make-range! "class.inner" @_start @_end)))
```
it works for me when you remove the duplicate `@_end`
